src/sage/rings/polynomial/q_integer_valued_polynomials.py

```diff
@@ -0,0 +1,1170 @@
+            return self.element_class(self, {NonNeg(0): ring(2),
+                                             NonNeg(2): ring(4)})
+
+        def _realization_name(self) -> str:
+            r"""
+            TESTS::
+
+                sage: F = QuantumValuedPolynomialRing(QQ).S()
+                sage: F._realization_name()
+                'shifted'
+            """
+            return "shifted"
+
+        def product_on_basis(self, n1, n2):
+            r"""
+            Return the product of basis elements ``n1`` and ``n2``.
+
+            INPUT:
+
+            - ``n1``, ``n2`` -- integers
+
+            EXAMPLES::
+
+                sage: A = QuantumValuedPolynomialRing(QQ).S()
+                sage: A.product_on_basis(0, 1)
+                S[1]
+            """
+            i = ZZ(n1)
+            j = ZZ(n2)
+            if j < i:
+                j, i = i, j
+            q = self.base_ring().gen()
+            return self._from_dict({i + j - k: (-1)**k
+                                    * q_binomial(i, k)
+                                    * q_binomial(i + j - k, i)
+                                    * q**(binomial(k, 2) - i * j)
+                                    for k in range(i + 1)})
+
+        def _from_binomial_basis(self, i):
+            """
+            Convert from the ``binomial(x,k)`` basis.
+
+            INPUT:
+
+            - ``i`` -- an integer
+
+            EXAMPLES::
+
+                sage: S = QuantumValuedPolynomialRing(ZZ).S()
+                sage: B = QuantumValuedPolynomialRing(ZZ).B()
+                sage: b = B.basis()
+                sage: S(b[3]+1)  # indirect doctest
+                -(q^-6-1)*S[0] + (q^-8+q^-7+q^-6)*S[1]
+                - (q^-9+q^-8+q^-7)*S[2] + (q^-9)*S[3]
+                sage: B(_)
+                B[0] + B[3]
+            """
+            i = ZZ(i)
+            R = self.base_ring()
+            q = self.base_ring().gen()
+            return self._from_dict({k: R((-1)**(i - k) * q_binomial(i, k))
+                                    * q**(-i**2 + binomial(i - k, 2))
+                                    for k in range(i + 1)})
+
+        def from_h_vector(self, hv):
+            """
+            Convert from some `h`-vector.
+
+            EXAMPLES::
+
+                sage: A = QuantumValuedPolynomialRing(ZZ).S()
+                sage: B = A.basis()
+                sage: ex = B[2] + B[3]
+                sage: A.from_h_vector(ex.h_vector())
+                S[2] + S[3]
+
+                sage: q = A.base_ring().gen()
+                sage: ex = B[2] + q*B[3]
+                sage: A.from_h_vector(ex.h_vector())
+                S[2] + q*S[3]
+            """
+            B = self.basis()
+            ring = self.base_ring()
+            q = ring.gen()
+            d = len(hv) - 1
+            m = matrix(ring, d + 1, d + 1,
+                       lambda j, i: (-1)**(d - j) * q_binomial(d - i, d - j, q) *
+                       q**(-d * (d - i) + binomial(d - j, 2)))
```

Comment:
Any reason why you aren't simply creating the lists of lists (or a single list) to pass to the constructor here?